### PR TITLE
Find `a` tags first and then parse them for `href`s

### PR DESCRIPTION
**Pull request by ryanscottaudio**
This addresses https://github.com/premailer/premailer/issues/404.

Simply making the `href` portion of the original regex optional didn't work for globbing reasons.

### CODE REVIEW

File: lib/premailer/html_to_plain_text.rb [hunk 1]
```diff
@@ -39,25 +39,24 @@ def convert_to_text(html, line_length = 65, from_charset = 'UTF-8')
...
+      ["'", '"'].each() do |quote_char|
+        match = /href=#{quote_char}(mailto:)?([^#{quote_char}]*)#{quote_char}/.match(s)
+        if !match.nil?
+          href = match[2]
+        end
       end
-    end
 
-    # links with single quotes
-    txt.gsub!(/<a\s[^\n]*?href=["'](mailto:)?([^']*)['][^>]*>(.*?)<\/a>/im) do |s|
```
Comment (grosser) [inline lib/premailer/html_to_plain_text.rb:54]: isn't the simpler solution to use `['"]` instead if `[']` ?

Comment (ryanscottaudio) [reply lib/premailer/html_to_plain_text.rb:54]: can you elaborate? we need to make sure the quotes on either side of the attr value match

Comment (grosser) [reply lib/premailer/html_to_plain_text.rb:54]: can we use `\1` ? 
```
>> "'hi'there" =~ /(['"])hi\1there/
=> 0
```

Comment (ryanscottaudio) [reply lib/premailer/html_to_plain_text.rb:54]: yup, that works

Comment (ryanscottaudio) [reply lib/premailer/html_to_plain_text.rb:54]: no, that breaks the tests

File: lib/premailer/html_to_plain_text.rb [hunk 2]
```diff
@@ -39,25 +39,24 @@ def convert_to_text(html, line_length = 65, from_charset = 'UTF-8')
...
-    # links with double quotes
-    txt.gsub!(/<a\s[^\n]*?href=["'](mailto:)?([^"]*)["][^>]*>(.*?)<\/a>/im) do |s|
-      if $3.empty?
-        ''
-      elsif $3.strip.downcase == $2.strip.downcase
-        $3.strip
-      else
-        $3.strip + ' ( ' + $2.strip + ' )'
+    # links
+    txt.gsub!(/<a[\s]+([^>]+)>((?:.(?!\<\/a\>))*.)<\/a>/im) do |s|
```
Comment (jteich) [inline lib/premailer/html_to_plain_text.rb:43]: Simpler, but logically equivalent: 

```
    txt.gsub!(/<a\s+([^>]+)>(.+?)<\/a>/im) do |s|
```

Note that the original version does not require any "content", but your updated version does. Is this change intentional? If not, then you can use:
```
    txt.gsub!(/<a\s+([^>]+)>(.*?)<\/a>/im) do |s|
```

File: lib/premailer/html_to_plain_text.rb [hunk 3]
```diff
@@ -39,25 +39,24 @@ def convert_to_text(html, line_length = 65, from_charset = 'UTF-8')
...
+      text = $2.strip
+      href = nil
+
+      ["'", '"'].each() do |quote_char|
+        match = /href=#{quote_char}(mailto:)?([^#{quote_char}]*)#{quote_char}/.match(s)
+        if !match.nil?
+          href = match[2]
+        end
       end
```
Comment (jteich) [inline lib/premailer/html_to_plain_text.rb:52]:
``` suggestion
      match = /href=(['"])(?:mailto:)?([^'"]+)\1/.match(s)
      if match
        href =match[1]
      end
```

Comment (ryanscottaudio) [reply lib/premailer/html_to_plain_text.rb:52]: i think you mean `match[2]`, but even that breaks the tests, as we need to have the quotation marks around the `href` value match each other (and we can have instances of the OTHER quotation mark inside the value)

Comment (jteich) [reply lib/premailer/html_to_plain_text.rb:52]: you're right. I fixed it:

``` suggestion
      match = /href=(['"])(?:mailto:)?(.+?)\1/.match(s)
      if match
        href =match[2]
      end
```

**Comment by ryanscottaudio** [comment]
@grosser how's this?

**Review by grosser — APPROVED** [review]
awesome!

**Comment by grosser** [comment]
1.15.0
